notify/receiver/receiver.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receiver
+
+import (
+	"github.com/go-kit/log"
+
+	commoncfg "github.com/prometheus/common/config"
+
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/notify/discord"
+	"github.com/prometheus/alertmanager/notify/email"
+	"github.com/prometheus/alertmanager/notify/msteams"
+	"github.com/prometheus/alertmanager/notify/opsgenie"
+	"github.com/prometheus/alertmanager/notify/pagerduty"
+	"github.com/prometheus/alertmanager/notify/pushover"
+	"github.com/prometheus/alertmanager/notify/slack"
+	"github.com/prometheus/alertmanager/notify/sns"
+	"github.com/prometheus/alertmanager/notify/telegram"
+	"github.com/prometheus/alertmanager/notify/victorops"
+	"github.com/prometheus/alertmanager/notify/webex"
+	"github.com/prometheus/alertmanager/notify/webhook"
+	"github.com/prometheus/alertmanager/notify/wechat"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+)
+
+type Wrapper func(string, notify.Notifier) notify.Notifier
+
+var NoWrap Wrapper
+
+// BuildReceiverIntegrations builds a list of integration notifiers off of a
+// receiver config.
+func BuildReceiverIntegrations(nc config.Receiver, tmpl *template.Template, logger log.Logger, wrap Wrapper, httpOpts ...commoncfg.HTTPClientOption) ([]notify.Integration, error) {
+	var (
+		errs         types.MultiError
+		integrations []notify.Integration
+		add          = func(name string, i int, rs notify.ResolvedSender, f func(l log.Logger) (notify.Notifier, error)) {
+			n, err := f(log.With(logger, "integration", name))
+			if err != nil {
+				errs.Add(err)
+				return
+			}
+			if wrap != nil {
+				n = wrap(name, n)
+			}
```

Comment:
I'd like to drop this abstraction - its only real use, is that the wrapper code can correlate Name->Notifier and do special things based on the notifier type. The notifiers cannot be disambiguated based on what is returned by this function.

A couple alternative approaches:
- Add a function `func Type() string` to the `notify.Notifier` interface, that lets code see what type a notifier is.
- Return a map of `type->Notifier` rather than an opaque slice of notifiers from this function.

The current approach is what `grafana/mimir` does: https://github.com/grafana/mimir/blob/7746f752c23fd5499fca59268d694a3345cfb29e/pkg/alertmanager/alertmanager.go#L466

Do any others have thoughts here?

Reply:
I like the first approach of modifying the Notifier interface - I think it's quite useful at the expense of a breaking change.

Reply:
Experimented with this idea a bit -

this is a deep change, for a PR that's trying to refactor something else. There's no common embedded object among the receivers, so we either need to introduce one, or add the method to each receiver individually. It's not complicated, but it touches a lot of code.

I also think the map approach might have merit and would like to experiment a bit.

I'm happy to do this, but perhaps as a separate PR that's not just a simple function export? I'll drop the `wrap()` bit from here, and we can figure out the best extension point as part of its own change.